apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/ActivationType.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.configuration;
+
+import co.elastic.apm.agent.impl.ElasticApmTracer;
+import co.elastic.apm.agent.impl.GlobalTracer;
+import co.elastic.apm.agent.util.PrivilegedActionUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.lang.management.ManagementFactory;
+import java.util.List;
+
+public enum ActivationType {
+    K8S_ATTACH, // https://github.com/elastic/apm-mutating-webhook
+    PROGRAMMATIC_SELF_ATTACH, // ElasticApmAttacher.attach();
+    JAVAAGENT_FLAG, // -javaagent:... command-line option used
+    AWS_LAMBDA_LAYER, // Only if installed by using layers (as other metadata already identifies lambda)
+    ENV_ATTACH, // JAVA_TOOL_OPTIONS env var used to specify -javaagent option
+    APM_AGENT_ATTACH_CLI, // 'java -jar apm-agent-attach-cli.jar ...' used
+    FLEET, // Fleet using 'java -jar apm-agent-attach-cli.jar ...' to attach (directly or through webhook)
+
+    AZURE_FUNCTIONS, //Not defined yet how this is identified
+    UNKNOWN;
+
+    private static final Logger logger = LogManager.getLogger(ActivationType.class);
+
+    public String toReferenceString() {
+        return toString().replace('_', '-').toLowerCase();
+    }
+
+    public static ActivationType findType() {
```

Comment:
See above - I think it should be called only once and not every time we serialize metadata

Reply:
Worst case overhead is 1 millisecond (in interpreted mode) and 2 additional objects each time the metadata is serialized (which I think is once every 5 seconds). Do we want to add a cache for that low an overhead? I'm asking rather than objecting

Reply:
This is not about overhead, it is conceptually wrong. As far as I understand, this is just like any other metadata information that we discover once and then use for the rest of the agent lifespan. And as such, it should be just another discovery step that is done by our `MetaData`, where it also needs to reside.
I wouldn't call it caching, it is part of the immutable discoverable information we collectively call `MetaData`.
TBH, the most proper treatment for the metadata is not only discover it once, but also serializing it once and forget about it, I just didn't get to that as it requires some refactoring in our serialization.

Reply:
If the right place conceptually to cache the metadata *is* at the serialization, then this is the wrong place to add another cache, especially since in that case this would be called just once.

Reply:
But I am not proposing to move it to the serialization in this PR. As I said, it will require a lot of refactoring because of how it is done now - the thing is that the metadata collection process involves some actions that we wanted to do in an unblocking way, asynchronously to other processes that may want to start reporting data, which requires the metadata to be ready. If you dive into it, you will see what I mean - it is doable, but possibly too difficult and risky to be worth the tiny improvement.
What you did, however, defies the metadata concept altogether as you invoke a discovery of immutable information every time we create a new request. So instead of adding another piece of info to this immutable data "cache", you propose to introduce a new way of getting metadata information.

Anyway, to conclude- I think that the activation method would be a very natural field addition to `co.elastic.apm.agent.impl.metadata.Agent` and that the most natural way to populate it would be when this object is created. 
If you follow the code and related flows and think otherwise- feel free to place elsewhere.

Reply:
Here you are asserting that metadata never changes. If, as we've discussed we could do, we enable the agent to be disabled so that it can be upgraded in place, then feasibly the activation could start as `JAVAAGENT_FLAG` and then later become `APM_AGENT_ATTACH_CLI`. In this scenario caching the activation now has introduced a bug that only shows later when everyone has forgotten this is cached

Reply:
I am saying that metadata is immutable **_during the lifespan of an agent_**. If we ever add a hot-upgrade capability then this would be considered a new agent instance with newly created metadata. 
`co.elastic.apm.agent.impl.metadata.Agent#version` would change for sure and probably `co.elastic.apm.agent.impl.metadata.Agent#ephemeralId` and other metadata fields as well. And this is all irrelevant because our intention is to completely remove all agent classes, class loaders and threads and start over again with almost all new agent code, including new `MetaData` class with possibly different fields.

Reply:
cached now